### PR TITLE
LPS-132570 Duplicate error messages show up when resetting password

### PR DESCRIPTION
**Pull request by holatuwol**
## Steps to reproduce

1. Set up an SMTP server to listen to email requests, such as http://nilhcem.com/FakeSMTP/ 
2. Navigate to Control Panel > Instance Settings > search for "mail" and select Mail Sender
3. Update the email address to something not on the blacklist (I chose jb@liferay.com)
4. Navigate to Control Panel > Server Administration > Mail and set the SMTP server info to match the SMTP server you setup in step 1
5. Create a new user, and set their password to something you know
6. Open a new web browser, and click on sign in > forgot password, enter in the email address of the user you created, and request a password reset email
7. Enter the same password as the password you set it to in the earlier step.

**Expected**: The "Your request failed to complete" toast error message should only show once. Not multiple times.

**Actual**: You will get an error "your password cannot be the same as the old password", and two "Your request failed to complete" toast error messages show up at the bottom left corner.

## Initial analysis

This appears to have been caused by there being two portlets (site navigation menu and search bar) on this placeholder page, each one running the `liferay-theme:portlet-messages` tag library, which in turn invokes the `liferay-ui:error` with no key, causing the toast message Javascript to appear in the page multiple times.

Adding code to remember if it's already shown at least once resolves the issue, but if there's a different way that you suspect may resolve the issue, or if there's something you'd like me to investigate, please let me know.

### CODE REVIEW

**Comment by holatuwol** [comment]
ci:test:sf

File: util-taglib/src/com/liferay/taglib/ui/ErrorTag.java [hunk 1]
```diff
@@ -354,6 +355,20 @@ private boolean _isShowAlert() {
...
+
+				if (suppressDefaultToastMessage) {
+					return false;
+				}
+
+				SessionErrors.add(
+					httpServletRequest,
+					"liferay-ui:error:suppressDefaultToastMessage");
+			}
+
```
Comment (markocikos) [inline util-taglib/src/com/liferay/taglib/ui/ErrorTag.java:371]: @holatuwol This does not seem right. We are adding adding logic for a specific error to a general error class. The issue should ideally be fixed on portlet level, not on tag level.

An error thrown by a portlet should normally be handled only by that portlet. Why is this case different? Can we apply standard error behavior?

Comment (holatuwol) [reply util-taglib/src/com/liferay/taglib/ui/ErrorTag.java:371]: This code adds logic that only runs for non-specific errors (no key specified).

Comment (markocikos) [reply util-taglib/src/com/liferay/taglib/ui/ErrorTag.java:371]: Let me clarify my question. I am not referring to code in this PR, I am referring to error-handling code in site navigation menu and search bar. You said in the opening comment they are the ones throwing errors. Why are they displaying errors thrown by a different portlet? This is not standard behavior. Why is this case different? Can we apply standard error behavior in those two portlets?

Comment (holatuwol) [reply util-taglib/src/com/liferay/taglib/ui/ErrorTag.java]: The gist of it is, we can't actually know which portlet generated the error message because it wasn't a portlet that generated the error, but rather a password reset struts action that's not tied to any portlet.

This is effectively being done through `portlet_messages/page.jsp` which is run for every portlet (so it's not as though these portlets are specifically showing all errors, but rather than portlet_messages, which is run for every portlet, will show the errors).

The issue happens due to changes made in [LPS-99142](https://issues.liferay.com/browse/LPS-99142), and I can fix it by adding another check to only render that message if it's the current portlet, but then the fix is functionally equivalent to calling SessionErrors instead of MultiSessionErrors, which makes me wonder if this change introduces some kind of regression that I can't understand just by looking at the code.

As part of due diligence, I test and confirmed that the original issue reported in LPS-99142 does not recur with the new changes, but I don't know whether there is another edge case that I should test. This is why I feel like an ErrorTag check would be safer, but I've applied the changes to `portlet_messages/page.jsp` instead.

File: portal-web/docroot/html/taglib/theme/portlet_messages/page.jsp [hunk 2]
```diff
@@ -97,6 +97,6 @@ else if (group.isStagingGroup()) {
 <liferay-ui:success key="<%= portlet.getPortletId() + SessionMessages.KEY_SUFFIX_UPDATED_CONFIGURATION %>" message="you-have-successfully-updated-the-setup" />
 <liferay-ui:success key="<%= portlet.getPortletId() + SessionMessages.KEY_SUFFIX_UPDATED_PREFERENCES %>" message="you-have-successfully-updated-your-preferences" />
 
-<c:if test="<%= !MultiSessionErrors.contains(renderRequest, NoSuchLayoutException.class.getName()) && !MultiSessionMessages.contains(renderRequest, portlet.getPortletId() + SessionMessages.KEY_SUFFIX_HIDE_DEFAULT_ERROR_MESSAGE) %>">
+<c:if test="<%= Objects.equals(portlet.getPortletName(), PortalUtil.getPortletId(request)) && !MultiSessionErrors.contains(renderRequest, NoSuchLayoutException.class.getName()) && !MultiSessionMessages.contains(renderRequest, portlet.getPortletId() + SessionMessages.KEY_SUFFIX_HIDE_DEFAULT_ERROR_MESSAGE) %>">
```
Comment (markocikos) [inline portal-web/docroot/html/taglib/theme/portlet_messages/page.jsp:100]: @holatuwol I agree with you in https://github.com/liferay-frontend/liferay-portal/pull/1104#discussion_r648065582, this looks risky and with possible uncovered edge cases.

Maybe the way to go is to treat password exception the same way `NoSuchLayoutException` is handled here, something like 
```jsp
<c:if test="<%= !MultiSessionErrors.contains(renderRequest, UserPasswordException.MustNotBeEqualToCurrent.class.getName()) && ...
```
I tested this and it prevents all toast messages, leaving only the inline one. This to me seems like the correct behavior, rather then less worse showing of only one toast.

I am not sure if there is a simple way in Java to cover all subtypes of `UserPasswordException`.

What do you think?

Comment (holatuwol) [reply portal-web/docroot/html/taglib/theme/portlet_messages/page.jsp]: It's a little weird since everything is a string, rather than class names. However, since all of the subtypes of UserPasswordException are inside the class itself, we can add all the declared classes to the checks. I've pushed an update which uses that check instead.

File: portal-kernel/src/com/liferay/portal/kernel/servlet/MultiSessionErrors.java [hunk 3]
```diff
@@ -92,7 +94,13 @@ private static boolean _isEmpty(HttpServletRequest httpServletRequest) {
...
 
 	private static Class<?>[] _hideDefaultErrorMessageClasses = {
-		NoSuchLayoutException.class
+		NoSuchLayoutException.class, UserPasswordException.class
 	};
 
+	static {
+		_hideDefaultErrorMessageClasses = ArrayUtil.append(
+			_hideDefaultErrorMessageClasses,
+			UserPasswordException.class.getDeclaredClasses());
```
Comment (markocikos) [inline portal-kernel/src/com/liferay/portal/kernel/servlet/MultiSessionErrors.java:103]: `getDeclaredClasses` <-- Nice! 

Can we add `UserPasswordException.class.getDeclaredClasses()` inline on L97?

File: portal-kernel/src/com/liferay/portal/kernel/servlet/MultiSessionErrors.java [hunk 4]
```diff
@@ -63,6 +64,25 @@ public static boolean isEmpty(PortletRequest portletRequest) {
 		return false;
 	}
 
+	public static boolean isHideDefaultErrorMessage(
+		PortletRequest portletRequest, String portletId) {
+
+		String key =
+			portletId + SessionMessages.KEY_SUFFIX_HIDE_DEFAULT_ERROR_MESSAGE;
```
Comment (markocikos) [inline portal-kernel/src/com/liferay/portal/kernel/servlet/MultiSessionErrors.java:73]: Minor SF: There is no need for `key` var, this can be inline on L73.

**Review by markocikos — CHANGES_REQUESTED** [review]
@holatuwol Looks good to me! Only a few minor SF nits inline.

The fact that it's a change on `portal-kernel` makes me a bit nervous, it doesn't typically fall under frontend responsibilities :sweat_smile: But the change does look safe to me and I tested it to work correctly (see gif below)

Side-question: Is there a way to redeploy `portal-kernel`, without doing `ant all`?

After PR:
after

**Comment by holatuwol** [comment]
> Side-question: Is there a way to redeploy `portal-kernel`, without doing `ant all`?

Yes. Shut down the server, navigate to `portal-kernel`, and run `ant-deploy`.

File: portal-kernel/src/com/liferay/portal/kernel/servlet/MultiSessionErrors.java (hunk 3 above)
Comment (holatuwol) [reply portal-kernel/src/com/liferay/portal/kernel/servlet/MultiSessionErrors.java:102]: Yes, this is done, and I've made it static final.

File: portal-kernel/src/com/liferay/portal/kernel/servlet/MultiSessionErrors.java (hunk 4 above)
Comment (holatuwol) [reply portal-kernel/src/com/liferay/portal/kernel/servlet/MultiSessionErrors.java]: Got it, went ahead and inlined it.

**Comment by markocikos** [comment]
@holatuwol LGTM :+1: 

I believe we need to update semver before forwarding. I ran `ant-all` and it updated two values in `portal-kernel`. I'm not sure why, but `gradle baseline` on `portal-kernel` did not update anything.

**Comment by markocikos** [comment]
Also, since this is a bit risky PR, I'd like to put on hold until the GA release. That should hopefully not be long.

**Comment by holatuwol** [comment]
We'd need to wait until after the GA release to redo the semver, because portal-kernel been constantly updated over the last few days.

**Review by markocikos — APPROVED** [review]

**Comment by markocikos** [comment]
@holatuwol GA2 was released. Please update semver, so we can forward away.

**Comment by holatuwol** [comment]
Added semver. Looks like there were changes that bumped the portal-kernel version yesterday already, so we only need to update the packageinfo.

**Comment by markocikos** [comment]
ci:forward
